### PR TITLE
Set ConnectionExpired Listener to MetadataClientDriver in AR

### PR DESCRIPTION
**Pull request by reddycharan**
Descriptions of the changes in this PR:

- add setConnectionExpiredListener method to MetadataClientDriver interface.
- add listener to shutdown AR in the case of metadata connection expiry

### Motivation

This commit - 4f0d2a195bd9be3788876b47813cee1440cf005c, has removed the shutdown logic in AutoRecoveryMain incase of ZK client session expiry, with the following reason "which doesn't make any sense for current retryable zookeeper". But if the ZK session has expired then it is not completely correct to let AutoRecoveryMain to continue to run in that state.

### CODE REVIEW

**Comment by reddycharan** [comment]
@athanatos created WIP PR for this concern.

**Comment by sijie** [comment]
I share same question as @eolivelli ; the underlying zookeeper client has infinite retry logic. if session expired on zookeeper, when auditor retries later, it will eventually succeed when established connections to zookeeper.

do you guys experience any problems on this? or do you have any tests that reproduce this issue?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/MetadataClientDriver.java [hunk 1]
```diff
@@ -84,4 +84,23 @@ LedgerManagerFactory getLedgerManagerFactory()
     @Override
     void close();
 
+    /**
+     * Listener on listening the metadata client connection expiry event.
+     */
+    @FunctionalInterface
```
[the remark this answers is not in the document]
Comment (athanatos) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/MetadataClientDriver.java]: I actually think that a notification that events may have been dropped is a fairly general concept.  SessionStateListener might be better though.

**Comment by reddycharan** [comment]
@eolivelli @sijie 

**Issue Summary:** I added the test case validating the need of this fix. As you can notice in this test case, with out this fix, when AR (with Auditor)’s ZK session expires, then as expected other AR will become Auditor and start running Auditor. But unexpectedly the previous Auditor also continue to run.

If AR is **not Auditor**, then because of the following ElectionWatcher - https://github.com/apache/bookkeeper/blob/master/bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/AuditorElector.java#L224 in the case of ZK session expiry, AuditorElector will shutdown and hence AutoRecoveryMain will shutdown because of https://github.com/apache/bookkeeper/blob/master/bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/AutoRecoveryMain.java#L230.

If AR is **Auditor**, then we are not setting ElectionWatcher in AuditorElector and hence above mentioned path wouldn’t execute. So AR and its Auditor will continue to run.

But the ‘myVote’ ephemeral node of the Auditor’s AR would be deleted in the case of ZK session expiry and hence some other AR will get this event and would become Auditor.

So effectively two AR’s are running Auditors now :( 

In summary just like what we are doing in the case of non-Auditor AR, we should shutdown the AR in the Auditor case as well. 

To be honest, I never fully grasped the semantics, guarantees and invariants of ZooKeeperClient wrapper class, especially regarding ’connectRetryPolicy’. @athanatos  has some strong concerns with how things (watchers, ephemeral nodes) are handled in ZooKeeperClient when connection is reestablished. Also, frequently I see people making misassumptions with what ZooKeeperClient offers.

@jvrao @codingwangqi fyi

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/MetadataClientDriver.java [hunk 2]
```diff
@@ -84,4 +84,23 @@ LedgerManagerFactory getLedgerManagerFactory()
...
+    interface ConnectionExpiredListener {
+
+        /**
+         * Signal when connection is expired.
+         */
+        void onConnectionExpired();
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/MetadataClientDriver.java]: I would call this simply 'onFatalError' or something that is simply telling that service should suicide.

Session expiration, and maybe session, is a ZK concept.

**Comment by eolivelli** [comment]
Okay you convinced me.

Let's work on naming and then ship this fix

**Comment by athanatos** [comment]
Generally speaking, the transparent reconnect logic in ZooKeeperClient is dangerous in two ways:
1) If the user isn't aware of what's going on, ephemeral nodes and watches can silently disappear.
2) Writes are generally non-idempotent, so retrying them can result in erroneous error conditions, as with node creation returning an EEXIST when it actually succeeded.

It's easy to reproduce errors like 2 by introducing intermittent packet drops while creating ledgers.

**Comment by sijie** [comment]
@athanatos : yes. the users of ZooKeeperClient should be aware of the two points you listed.

**Comment by reddycharan** [comment]
@sijie can you comment on validity of this fix and feedback on interface method name

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/AutoRecoveryMain.java [hunk 3]
```diff
@@ -91,6 +92,11 @@ public AutoRecoveryMain(ServerConfiguration conf, StatsLogger statsLogger)
             CompatibilityException {
         this.conf = conf;
         this.bkc = Auditor.createBookKeeperClient(conf);
+        MetadataClientDriver metadataClientDriver = bkc.getMetadataClientDriver();
+        metadataClientDriver.setConnectionExpiredListener(() -> {
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/AutoRecoveryMain.java]: I think a cleaner solution is AuditorElector handling this session expires. because AuditorElector already has this session expires watcher to shutdown elector - https://github.com/apache/bookkeeper/blob/master/bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/AuditorElector.java#L224 so all the leader election logic are all in one place, rather than spreading across multiple places. and it seems to be working better with #1701 (https://github.com/apache/bookkeeper/pull/1701/files#diff-7d01af204bb0a30c0193b46c737f8cbfR49)

However I am fine with this change as it is working well with the zookeeper concepts. We can go with this change for now, once the change #1701 is merged, we can considering moving this session expires logic in to the AuditorElector, which will provide a much cleaner abstraction.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/MetadataClientDriver.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (sijie) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/MetadataClientDriver.java]: IMO SessionStateListener is a better name than ConnectionExpiredListener

[the remark this answers is not in the document]
Comment (eolivelli) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/MetadataClientDriver.java]: +1 for SessionStateListener

File: metadata-drivers/etcd/src/main/java/org/apache/bookkeeper/metadata/etcd/EtcdMetadataClientDriver.java [hunk 4]
```diff
@@ -72,4 +72,12 @@ public synchronized void close() {
         }
         super.close();
     }
+
+    @Override
+    public void setSessionStateListener(SessionStateListener sessionStateListener) {
+        /*
+         * TODO: EtcdMetadataClientDriver has to implement this method.
+         */
+        throw new UnsupportedOperationException();
```
Comment (reddycharan) [inline metadata-drivers/etcd/src/main/java/org/apache/bookkeeper/metadata/etcd/EtcdMetadataClientDriver.java]: @codingwangqi im just throwing exception for now. As part of metadataclientdriver/etcd changes can you please take care of this.

@sijie

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/MetadataClientDriver.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (reddycharan) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/MetadataClientDriver.java]: changed interface/method name

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/AutoRecoveryMain.java (hunk 3 above)
Comment (codingwangqi) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/AutoRecoveryMain.java]: @sijie thank you for pointing this out. yes, I think session expires is an implementation specific behavior, I will incorporate your proposal into #1701 once this change is merged.

File: metadata-drivers/etcd/src/main/java/org/apache/bookkeeper/metadata/etcd/EtcdMetadataClientDriver.java (hunk 4 above)
Comment (codingwangqi) [reply metadata-drivers/etcd/src/main/java/org/apache/bookkeeper/metadata/etcd/EtcdMetadataClientDriver.java]: @reddycharan thank you for letting me know about this. I will take care of it.

**Comment by reddycharan** [comment]
@eolivelli need your final approval

**Comment by reddycharan** [comment]
@sijie i think you approved this. So it should be good to go.

**Comment by eolivelli** [comment]
#shipit
